html/e2etest.js

```diff
@@ -0,0 +1,654 @@
+// the page, meaning that, ideally, we'll be the only one able to decrypt
+// it: Janus will not have access to the media. In a scenario involving
+// multiple participants (e.g., a video call or conference), to use these
+// same functions you'll want to distribute the secret among all participants
+// using an out of band mechanism, so that they can encrypt and decrypt
+// the media and keep Janus out of the loop. Notice that these functions
+// are just an example, and are basically exact copies from this demo:
+// 		https://github.com/webrtc/samples/blob/gh-pages/src/content/peerconnection/endtoend-encryption/js/main.js
+// That said, you're free (or actually, encouraged) to try and use other
+// custom, and more advanced, trasform functions as well (e.g., SFrames).
+var currentCryptoKey = null;
+var currentKeyIdentifier = 0;
+function promptCryptoKey() {
+	bootbox.prompt("Insert crypto key (e.g., 'mysecret') to use", function(result) {
+		if(!result || result === "") {
+			promptCryptoKey();
+			return;
+		}
+		// Take note of the secret
+		currentCryptoKey = result;
+		currentKeyIdentifier++;
+		// Negotiate the WebRTC PeerConnection, now (clone of EchoTest demo code)
+		Janus.debug("Trying a createOffer too (audio/video sendrecv)");
+		echotest.createOffer(
+			{
+				// No media provided: by default, it's sendrecv for audio and video
+				media: { data: true },	// Let's negotiate data channels as well
+				// If you want to test simulcasting (Chrome and Firefox only), then
+				// pass a ?simulcast=true when opening this demo page: it will turn
+				// the following 'simulcast' property to pass to janus.js to true.
+				simulcast: doSimulcast,
+				simulcast2: doSimulcast2,
+				// Since we want to use Insertable Streams,
+				// we specify the transform functions to use
+				senderTransform: senderTransform,
+				receiverTransform: receiverTransform,
+				success: function(jsep) {
+					Janus.debug("Got SDP!", jsep);
+					var body = { audio: true, video: true, data: true };
+					// We can try and force a specific codec, by telling the plugin what we'd prefer
+					// For simplicity, you can set it via a query string (e.g., ?vcodec=vp9)
+					if(acodec)
+						body["audiocodec"] = acodec;
+					if(vcodec)
+						body["videocodec"] = vcodec;
+					echotest.send({ message: body, jsep: jsep });
+				},
+				error: function(error) {
+					Janus.error("WebRTC error:", error);
+					bootbox.alert("WebRTC error... " + error.message);
+				}
+			});
+	});
+}
+var senderTransform = new TransformStream({
+	start() {
+		// Called on startup.
+		console.log("[Sender transform] Startup");
+	},
+	transform(chunk, controller) {
+		//~ console.log("[Sender transform] Transforming");
+		// Copy of the above mentioned demo's encodeFunction()
+		const view = new DataView(chunk.data);
+		const newData = new ArrayBuffer(chunk.data.byteLength + 5);
+		const newView = new DataView(newData);
+		for(let i=0; i<10; ++i) {
```

Comment:
you might want to update that one, in particular for audio there are three-byte (and even 0-byte) frames.

Reply:
Thanks, I think I copied the functions a couple of weeks ago so it's very likely I'm using outdated examples. As soon as audio is supported in my browser I'll update and test that too.